feat: Projection pushdown for querier -> ingester for rpc queries

NGA-TRAN opened the pull request:
Closes #5741

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

NGA-TRAN commented on iox_query/src/test.rs:
```diff
@@ -108,18 +108,50 @@ impl QueryDatabase for TestDatabase {
+            .collect::<Vec<_>>();
+
+        let chunks = futures::future::join_all(chunks).await;
+        Ok(chunks)
```
@tustvold : here is the function that hits compile problem

NGA-TRAN replied:
Thanks @alamb who has found the issue

alamb commented:
If I applied this diff I got the code to compile. Not sure if that is correct or not

```diff
 diff --git a/iox_query/src/test.rs b/iox_query/src/test.rs
index ee4124fbb..2202c76a5 100644
--- a/iox_query/src/test.rs
+++ b/iox_query/src/test.rs
@@ -117,13 +117,12 @@ impl QueryDatabase for TestDatabase {
         // todo: apply projction to chunks
         println!("==== projection: {:?}", projection);
 
-        let partitions = self.partitions.lock();
+        let partitions = self.partitions.lock().clone();
         let chunks = partitions
             .values()
             .flat_map(|x| x.values())
             .filter(|x| x.table_name == table_name)
-            .map(|x| Arc::clone(x) as Arc<dyn QueryChunk>)
-            .collect::<Vec<_>>();
+            .map(|x| Arc::clone(x) as Arc<dyn QueryChunk>);
 
         let chunks = chunks
             .into_iter()
@@ -133,20 +132,19 @@ impl QueryDatabase for TestDatabase {
                 let cols = cols.iter().map(|c| c.as_str()).collect::<Vec<_>>();
                 let selection = Selection::Some(&cols);
 
-                let mut batches: Vec<RecordBatch> = vec![];
                 let mut stream = c
                     .read_filter(IOxSessionContext::with_testing(), predicate, selection)
                     .expect("Error in read_filter");
 
-                let mut new_chunk = Arc::new(TestChunk::new(c.table_name()));
+                let mut new_chunk = TestChunk::new(c.table_name());
                 while let Some(b) = stream.next().await {
                     let b = b.expect("Error in stream");
                     new_chunk.table_data.push(Arc::new(b));
                 }
                 assert!(!new_chunk.table_data.is_empty(), "Error: empty batches");
                 let new_schema = Schema::try_from(new_chunk.table_data[0].schema()).unwrap();
-                new_chunk.add_schema_to_table(new_schema, true, None);
-                new_chunk
+                let new_chunk = new_chunk.add_schema_to_table(new_schema, true, None);
+                Arc::new(new_chunk) as _
             })
             .collect::<Vec<_>>();
```

NGA-TRAN commented on iox_query/src/test.rs:
```diff
@@ -938,7 +971,14 @@ impl QueryChunk for TestChunk {
+                self.schema().as_arrow(),
+            )))
+        } else {
+            Ok(stream_from_batches(batches))
```
The `stream_from_batches` only works with non-empty batches. Need to add the empty stream to have all tests that now need to invoke this function work

crepererum replied:
Then `stream_from_batches` should be fixed (it probably needs an additional `schema` parameter for that).

NGA-TRAN commented on iox_query/src/frontend/influxrpc.rs:
```diff
@@ -1314,6 +1334,7 @@ impl InfluxRpcPlanner {
 /// Stream of chunks for table predicates.
 fn table_chunk_stream<'a>(
```
The main implementation is in this function. It has a new parameter `all_cols` and if true, nothing is pushed down. When it is false, only columns in the `rpc_predicate` will be read. Since we always read PK( tags and time) for deduplication, we only need to extract columns of the rpc_predicate's `exprs` and `value_expr`. 

1. Queries that has projection pushed down in this PR: `table_name`, `tag_keys`, `tag_values`
2. Queries I know that we cannot push anything down:  `field_columns`, `read_filter`
3. I am not sure about these 2 and still keep reading all columns for them are: `read_group` and `read_window_aggregate`.  @alamb : can you tell me whether we can eliminate some columns in these cases?

crepererum replied:
Why can we not push down projections for `read_filter`? This seems to be the major use case (all the cases you list under point 1 are metadata queries and ideally we would remove payload scan for these entirely at some point, if politics allows it).

crepererum replied:
Also, the parameter `all_cols` should probably be documented.

NGA-TRAN reviewed (commented):
@alamb and @crepererum: the PR is ready for review. I have a couple of questions whether we should have projection pushdown for `read_group` or `read_window_aggregate`

crepererum commented on iox_query/src/frontend/influxrpc.rs:
```diff
@@ -1324,9 +1345,62 @@ fn table_chunk_stream<'a>(
+                // Build projection that includes all columns in the predicate's exprs and val_expr
+                // Note that since PK cols (tags and time) will be added in later steps, inside the function
+                // ingester_partitions_inner,  we don't need to consider them here.
+                let mut exprs_columns = StdHashSet::new();
```
I think you can use the `hashbrown` hashset here as well. It should provide you all the features you need (IIRC its interface is mostly a superset of the std variant).

NGA-TRAN replied:
The `exprs_columns` is the mutable parameter of DF's `eprlist_to_columns` and it is defined as std::HashSet

crepererum commented on iox_query/src/frontend/influxrpc.rs:
```diff
@@ -1324,9 +1345,62 @@ fn table_chunk_stream<'a>(
+                None
+            } else {
+                // Build projection that includes all columns in the predicate's exprs and val_expr
+                // Note that since PK cols (tags and time) will be added in later steps, inside the function
+                // ingester_partitions_inner,  we don't need to consider them here.
```
I understand this explanation but the argument is extremely fragile. You argue with a currently existing, concrete implementation of an abstract interface that is used here. Instead, you should document in `QueryDatabase::chunks` that the returned data ALWAYS must includes PK columns, not matter if the projection removes them or not.

NGA-TRAN replied:
Done

crepererum commented on iox_query/src/frontend/influxrpc.rs:
```diff
@@ -1324,9 +1345,62 @@ fn table_chunk_stream<'a>(
+                // convert the column names into their corresponding indexes in the schema
+                let projection = {
+                    if expr_cols_result.is_err() || val_exprs_cols_result.is_err() {
+                        // cannot determine columns, use all columns
```
please add a debug log to this "read all columns" fallback.

crepererum commented on iox_query/src/frontend/influxrpc.rs:
```diff
@@ -1324,9 +1345,62 @@ fn table_chunk_stream<'a>(
+                                    Some(cols.into_iter().flatten().collect::<Vec<_>>())
+                                }
+                            }
+                            _ => None,
```
IIRC the `table_schema` can only be `None` if the table doesn't exist. I guess you could exit (for this one table, not for all) early in this case and don't return any chunks at all.

NGA-TRAN replied:
It works. Since we return projection here, I make the projection None if table does not exist

crepererum commented on iox_query/src/frontend/influxrpc.rs:
```diff
@@ -1324,9 +1345,62 @@ fn table_chunk_stream<'a>(
+                                    .collect::<Vec<_>>();
+
+                                if cols.contains(&None) || cols.is_empty() {
+                                    // At least one column has no matching index, we do not know which
```
As above: please add a debug log (which should also say which case -- contains None or cols are empty -- was hit).

NGA-TRAN replied:
Done

crepererum commented on iox_query/src/frontend/influxrpc.rs:
```diff
@@ -1324,9 +1345,62 @@ fn table_chunk_stream<'a>(
 
             let database = Arc::clone(&database);
 
+            let projection = if all_cols {
```
Maybe this whole block can be moved into a new function? This would also help testing because you could test the "build a projection" code without building the entire stream (i.e. calling `table_chunk_stream`). Looking at your test code, I think this would be a net win.

crepererum commented on iox_query/src/test.rs:
```diff
@@ -108,18 +108,56 @@ impl QueryDatabase for TestDatabase {
+        projection: &Option<Vec<usize>>,
         _ctx: IOxSessionContext,
     ) -> Result<Vec<Arc<dyn QueryChunk>>, QueryDatabaseError> {
+        // return Err("Sugar we are going down".into());
```
what's that?

NGA-TRAN replied:
Good catch. It was a simulation for a failed test that I fixed later

crepererum commented on iox_query/src/frontend/influxrpc.rs:
```diff
@@ -1765,12 +1840,285 @@ mod tests {
+        let predicate = Predicate::new().with_expr(expr);
+        let table_predicates = vec![("h2o".to_string(), predicate)];
+
+        let result = table_chunk_stream(test_db, true /* all cols */, &table_predicates, &ctx)
```
If you find yourself documenting your boolean flags, maybe this is a warning sign and you shall use an enum instead :wink:

crepererum commented:
@alamb if you review this, hiding whitespaces in the resulting diff helps a lot (since a large chunk of the changes is rustfmt re-arranging the code).

alamb reviewed (approved):
I found whitespace blind diff https://github.com/influxdata/influxdb_iox/pull/5782/files?w=1 much easer to see what had changed

Another test I think would add value is a basic "is this plumbing hooked up" type test -- that runs a grpc query and then validates that not all columns were fetched. 

In general there are some structural changes that I think would help with readability, and I have some concerns about the specifics of implementation. However, I feel very good about the coverage in `query_tests` so the fact that this PR passes means I feel confident in approving

This is a nice improvement

NGA-TRAN reviewed (commented):
Very helpful comments @crepererum and @alamb . Let me address them in this PR

NGA-TRAN commented:
I convert this to draft because even though I have addressed the main comments, there are are few minor things I need to address tomorrow

NGA-TRAN commented on iox_query/src/frontend/influxrpc.rs:
```diff
@@ -1314,6 +1334,7 @@ impl InfluxRpcPlanner {
 /// Stream of chunks for table predicates.
 fn table_chunk_stream<'a>(
     database: Arc<dyn QueryDatabase>,
+    all_cols: bool,
```
@alamb : I have used `field_columns` per your suggestion to filter more data. However, I still need extra parameter in case this field_coumns is empty.  I renamed `all_cols` to `need_fields` for this purpose. More specifically, if need_fields is false (in case we only need to return tags or table name), we will do nothing if field_coumns is empty. When need_fields is true (in the case of field_columns, read_filter, ...) and field_coumns is empty, we have to read all field columns

NGA-TRAN commented:
@alamb and @crepererum : I have addressed all your comments and have a question. I will merge this shortly. If there are extra stuff you think we need to do, I will do it in next PR
